src/core/search/block_list.cc

```diff
@@ -6,48 +6,52 @@ using namespace std;
 
 SplitResult Split(BlockList<SortedVector<std::pair<DocId, double>>>&& block_list) {
   using Entry = std::pair<DocId, double>;
-
-  const size_t initial_size = block_list.Size();
   DCHECK(!block_list.Empty());
 
-  // Collect all entries from the blocks into a single vector
-  // We copy the block and clear it on the fly to save memory
-  std::vector<std::pair<DocId, double>> all_entries;
-  for (auto& block : block_list.blocks_) {
-    all_entries.reserve(all_entries.size() +
-                        block.Size());  // Reserve space for all entries in the block
-    all_entries.insert(all_entries.end(), block.begin(), block.end());
-    block.Clear();  // At the same time clear the block to save memory
-  }
-  block_list.Clear();
+  const size_t elements_count = block_list.Size();
 
-  DCHECK(all_entries.size() == initial_size);
+  using EntryIndex = std::pair<size_t /*block id*/, size_t /*entry id*/>;
+  std::vector<EntryIndex> entries_indexes(elements_count);
+  size_t index = 0;
+  for (size_t i = 0; i < block_list.blocks_.size(); ++i) {
+    auto& block = block_list.blocks_[i];
+    for (size_t j = 0; j < block.Size(); ++j) {
+      entries_indexes[index++] = {i, j};
+    }
+  }
 
-  std::nth_element(all_entries.begin(), all_entries.begin() + initial_size / 2, all_entries.end(),
-                   [](const Entry& l, const Entry& r) { return l.second < r.second; });
+  std::nth_element(entries_indexes.begin(), entries_indexes.begin() + elements_count / 2,
+                   entries_indexes.end(), [&](const EntryIndex& l, const EntryIndex& r) {
+                     return block_list.blocks_[l.first][l.second].second <
+                            block_list.blocks_[r.first][r.second].second;
+                   });
 
-  double median_value = all_entries[initial_size / 2].second;
+  const EntryIndex median_index = entries_indexes[elements_count / 2];
+  double median_value = block_list.blocks_[median_index.first][median_index.second].second;
 
   BlockList<SortedVector<Entry>> left(block_list.blocks_.get_allocator().resource(),
                                       block_list.block_size_);
   BlockList<SortedVector<Entry>> right(block_list.blocks_.get_allocator().resource(),
                                        block_list.block_size_);
   absl::InlinedVector<Entry, 1> median_entries;
 
+  left.ReserveBlocks(block_list.blocks_.size() / 2 + 1);
```

Comment:
don't you over reserve here?
say if `block_list.blocks_.size()` is 4, you reserve 3 in each list?